sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Metrics/ExecutableLinesCalculator.cs

```diff
@@ -0,0 +1,111 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2017 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Collections.Generic;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using SonarAnalyzer.Helpers;
+
+namespace SonarAnalyzer.Metrics.CSharp
+{
+    public class ExecutableLinesCalculator
+    {
+        public int CountLines(SyntaxTree tree)
+        {
+            var nodesToVisit = new Queue<SyntaxNode>(tree.GetRoot().ChildNodes());

```

Comment:
Not sure to understand why you take only top level nodes and you will take sub level ones after?

Reply:
Oh yes, I can use DescendantNodes instead.

Reply:
At some point we will need to make some performance test. @dbolkensteyn told me it was recommended by roslyn team not to use methods querying children and then using Linq but to rather subscribe to the different levels we are interested in.

Reply:
Interesting!

I've run the test (with walker and overriding `DefaultVisit`). The walker was quicker by 30ms, with total time of about 0.5s during ITs. That's probably within a margin of error.